src/ameba/rule/style/redundant_self.cr

```diff
@@ -0,0 +1,239 @@
+require "compiler/crystal/syntax/token"
+
+module Ameba::Rule::Style
+  # A rule that disallows redundant uses of `self`.
+  #
+  # This is considered bad:
+  #
+  # ```
+  # class Greeter
+  #   getter name : String
+  #
+  #   def self.init
+  #     self.new("Crystal").greet
+  #   end
+  #
+  #   def initialize(@name)
+  #   end
+  #
+  #   def greet
+  #     puts "Hello, my name is #{self.name}"
+  #   end
+  #
+  #   self.init
+  # end
+  # ```
+  #
+  # And needs to be written as:
+  #
+  # ```
+  # class Greeter
+  #   getter name : String
+  #
+  #   def self.init
+  #     new("Crystal").greet
+  #   end
+  #
+  #   def initialize(@name)
+  #   end
+  #
+  #   def greet
+  #     puts "Hello, my name is #{name}"
+  #   end
+  #
+  #   init
+  # end
+  # ```
+  #
+  # YAML configuration example:
+  #
+  # ```
+  # Style/RedundantSelf:
+  #   Enabled: true
+  #   AllowedMethodNames:
+  #     - in?
+  #     - inspect
+  #     - not_nil!
+  # ```
+  class RedundantSelf < Base
+    include AST::Util
+
+    properties do
+      since_version "1.7.0"
+      description "Disallows redundant uses of `self`"
+
+      allowed_method_names %w[in? inspect not_nil!]
+    end
+
+    MSG = "Redundant `self` detected"
+
+    CRYSTAL_KEYWORDS = Crystal::Keyword.values.map(&.to_s)
+
+    def test(source)
+      ScopeCallsWithSelfReceiverVisitor.new self, source
+    end
+
+    def test(source, node : Crystal::Call, scope : AST::Scope)
+      return unless (obj = node.obj).is_a?(Crystal::Var)
+
+      name = node.name
+
+      return if name.in?(CRYSTAL_KEYWORDS)
+      return if name.in?(allowed_method_names)
+      return if name.ends_with?('=')
+      return if name.chars.none?(&.alphanumeric?)
+
+      vars = Set(String).new
+
+      while scope
+        scope.arguments.each do |arg|
+          vars << arg.name
+        end
+        scope.variables.each do |var|
+          var.assignments.each do |assign|
+            vars << assign.variable.name
+          end
+        end
+        scope = scope.outer_scope
+      end
+
+      return if name.in?(vars)
+
+      if location = obj.location
+        issue_for obj, MSG do |corrector|
+          corrector.remove(location, location.adjust(column_number: {{ "self".size }}))
+        end
+      else
+        issue_for obj, MSG
+      end
+    end
+
+    private class ScopeCallsWithSelfReceiverVisitor < Crystal::Visitor
```

Comment:
There's a lot of duplication with `AST::ScopeVisitor`, thus an opportunity for a refactor.

Reply:
I recommend maintaining the original file structure by relocating this component to the visitors directory. Although it may seem that this class is currently used only by the existing rule, similar rules may leverage it in the future.

Reply:
This one is pretty specific, that's why I would like to keep it next to the rule. In case it would be needed we can always move it.

Reply:
Actually, I've thought this over and I've moved it aside, since it helps with the testing of the visitor class itself, apart of file organization.